src/librustc_mir/transform/mod.rs

```diff
@@ -260,6 +261,8 @@ fn run_optimization_passes<'tcx>(
         &uniform_array_move_out::RestoreSubsliceArrayMoveOut,
         &inline::Inline,
 
+        // State transform requires that predecessors have been predefined
+        &ensure_predecessors_cache::EnsurePredecessorsCache::new("pre-state-transform"),
```

Comment:
Can't the passes, having mutable access to the Body generate this themselves?

Reply:
They could but I would have to convert `&Body` to `&mut Body` in a few places. Some of the code that uses the `&Body` even does so explicitly, e.g. https://github.com/rust-lang/rust/blob/master/src/librustc_mir/borrow_check/mod.rs#L155. Do you think that makes more sense?

Reply:
Hm... The `MirPass` could call `ensure_predecessors` right in the start instead of as-needed.

Ok, with this argument we could just nuke `EnsurePredecessorsCache` entirely and call `ensure_predecessors` after the mir pass pipeline or in the beginning of any pass that needs it.

So... Either remove `EnsurePredecessorsCache` or keep the PR as is, I'm fine with either scheme, so your call.

Reply:
Turns out this specific one was unneeded. I do like having the pass in other runs (where the run is just after the body is created), so I'd prefer to leave those there. It makes the ensure a bit more explicit, whereas the others I feel need a detailed comment to explain their need.